shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -338,11 +339,13 @@ private void dispatchMessageToQueue(
   @Override
   public void handleMessageFromDart(
       @NonNull String channel, @Nullable ByteBuffer message, int replyId, long messageData) {
-    // Called from the ui thread.
+    // Called from any thread.
     Log.v(TAG, "Received message from Dart over channel '" + channel + "'");
 
     HandlerInfo handlerInfo;
     boolean messageDeferred;
+    // TODO(gaaclarke): This can potentially be a bottleneck and could replaced
+    // with a read/write lock.
```

Comment:
file a bug

Reply:
I switched it to a regular comment if you don't mind.  I'm not sure I want to commit to the idea that it is something we need to do, but I don't want to lose the idea that it was something we could do.

Reply:
I think it'd be fine to file a bug either way to add a benchmark for this and figure out if/how much a different lock would help.  But non-blocking for landing this.

Reply:
https://github.com/flutter/flutter/issues/111489